d2/src/main/java/com/linkedin/d2/balancer/strategies/relative/RelativeStateUpdater.java

```diff
@@ -0,0 +1,390 @@
+      latencySum += latestCallStats.getCallTimeStats().getAverage() * callCount;
+      outstandingLatencySum += latestCallStats.getOutstandingStartTimeAvg() * outstandingCallCount;
+      callCountSum += callCount;
+      outstandingCallCountSum += outstandingCallCount;
+    }
+
+    return callCountSum + outstandingCallCountSum == 0
+        ? 0
+        : (latencySum + outstandingLatencySum) / (callCountSum + outstandingCallCountSum);
+  }
+
+  private static long getAvgHostLatency(CallTracker.CallStats callStats)
+  {
+    double avgLatency = callStats.getCallTimeStats().getAverage();
+    long avgOutstandingLatency = callStats.getOutstandingStartTimeAvg();
+    int callCount = callStats.getCallCount();
+    int outstandingCallCount = callStats.getOutstandingCount();
+    return callCount + outstandingCallCount == 0
+        ? 0
+        : Math.round(avgLatency * (callCount / (callCount + outstandingCallCount))
+            + avgOutstandingLatency * (outstandingCallCount / (callCount + outstandingCallCount)));
+  }
+
+  /**
+   * Identify if a client is unhealthy
+   */
+  private boolean isUnhealthy(TrackerClientState trackerClientState, long avgClusterLatency,
+      int callCount, long latency, double errorRate)
+  {
+    return callCount >= trackerClientState.getAdjustedMinCallCount()
+        && (latency >= avgClusterLatency * _relativeStrategyProperties.getRelativeLatencyHighThresholdFactor()
+        || errorRate >= _relativeStrategyProperties.getHighErrorRate());
+  }
+
+  /**
+   * Identify if a client is healthy
+   */
+  private boolean isHealthy(TrackerClientState trackerClientState, long avgClusterLatency,
+      int callCount, long latency, double errorRate)
+  {
+    return callCount >= trackerClientState.getAdjustedMinCallCount()
+        && (latency <= avgClusterLatency * _relativeStrategyProperties.getRelativeLatencyLowThresholdFactor()
+        || errorRate <= _relativeStrategyProperties.getLowErrorRate());
+  }
+
+  private void emitMonitorEvents(PartitionRelativeLoadBalancerState state)
+  {
+    state.getListeners().forEach(listener -> listener.onUpdate(state));
+  }
+
+  private static double getErrorRateByType(Map<ErrorType, Integer> errorTypeCounts, int callCount)
+  {
+    Integer connectExceptionCount = errorTypeCounts.getOrDefault(ErrorType.CONNECT_EXCEPTION, 0);
+    Integer closedChannelExceptionCount = errorTypeCounts.getOrDefault(ErrorType.CLOSED_CHANNEL_EXCEPTION, 0);
+    Integer serverErrorCount = errorTypeCounts.getOrDefault(ErrorType.SERVER_ERROR, 0);
+    Integer timeoutExceptionCount = errorTypeCounts.getOrDefault(ErrorType.TIMEOUT_EXCEPTION, 0);
+    return callCount == 0
+        ? 0
+        : (double) (connectExceptionCount + closedChannelExceptionCount + serverErrorCount + timeoutExceptionCount) / callCount;
+  }
+
+  private void initializePartition(Set<TrackerClient> trackerClients, int partitionId, long clusterGenerationId)
+  {
+    Lock lock = new ReentrantLock();
```

Comment:
concurrency concern

Reply:
I changed to use only one week in the StateUpdater, this will be safe, but may slow down the requests a little bit if multiple requests are sent to different partitions during initialization.